zerver/lib/users.py

```diff
@@ -860,7 +908,14 @@ def is_2fa_verified(user: UserProfile) -> bool:
 
 
 def get_users_with_access_to_real_email(user_profile: UserProfile) -> List[int]:
-    active_users = user_profile.realm.get_active_users()
+    if not user_access_restricted_in_realm(user_profile):
+        active_users = user_profile.realm.get_active_users()
+    else:
+        user_ids_who_can_access_user = get_user_ids_who_can_access_user(user_profile)
+        active_users = UserProfile.objects.filter(
```

Comment:
We could have just directly got user objects here instead of first getting IDs and then objects, but we use the same code which is used for getting accessible users for multiple users, so that makes it somewhat inefficient to query large number of `UserProfile` objects.

And this extra query should not be much problematic as we need user objects only when changing emails and email visibility setting.

Reply:
This makes sense to me.